Header Bar Style Updates

iamnickvolpe opened the pull request:
# Header Bar Style Updates
## Summary
This PR contains various styling updates to the header bar and surrounding UI. The goal of this update is to add clarity to the interface. 

- Rearranged buttons based on a general formula where the left side is pertaining to the project being viewed, and the right side is about going further, or doing more "advanced" things.
- Varied the styles of buttons based on their importance and behavior.
- Added hover states to all buttons.
- Modified the active states of buttons.
- Modified the styles of menus.
- Modified the style of snapshot notification.

## Images
### Initial State
<img width="1048" alt="screenshot 2017-11-12 02 11 11" src="https://user-images.githubusercontent.com/3432105/32696863-1b10ac58-c751-11e7-9ab3-5f4743da17d2.png">

### Logged in state
<img width="1046" alt="screenshot 2017-11-12 02 11 27" src="https://user-images.githubusercontent.com/3432105/32696877-4c99e10e-c751-11e7-93c4-f05bb53eb9f6.png">

### Button Hover State
<img width="131" alt="screenshot 2017-11-12 02 11 46" src="https://user-images.githubusercontent.com/3432105/32696879-599b5a54-c751-11e7-9da1-ea752d1b7cf0.png">

### Button Active State
<img width="108" alt="screenshot 2017-11-12 02 12 08" src="https://user-images.githubusercontent.com/3432105/32696882-6ed2d636-c751-11e7-9395-6a31a1214273.png">

### Libraries Menu Design
<img width="120" alt="screenshot 2017-11-12 02 12 24" src="https://user-images.githubusercontent.com/3432105/32696885-7d638cae-c751-11e7-8ff1-e43d4166f5da.png">

### Projects Menu Design
<img width="323" alt="screenshot 2017-11-12 02 12 35" src="https://user-images.githubusercontent.com/3432105/32696889-883dcd06-c751-11e7-8c5d-0400982b0f33.png">

### Screenshots Notification Design
<img width="1042" alt="screenshot 2017-11-12 02 13 09" src="https://user-images.githubusercontent.com/3432105/32696893-950e26ca-c751-11e7-85c5-9938ff006fd9.png">

iamnickvolpe commented:
@outoftime This new PR addresses your comments about the previous one (the padding around the logo and the line-height of the header bar itself). 

One thing I inadvertently did though was do away with the "sizing hack." (https://github.com/popcodeorg/popcode/pull/1077) I think that hack was interesting but isn't really that great for accessibility and sort of unpredictable, making it difficult to reign in the new styling.

2 suggestions there: 1, making the "zoom" button a dropdown menu, allowing for more zoom levels, and 2, a "presentation mode" where the header bar is just hidden from view.

outoftime commented:
@iamnickvolpe agree that the zoom-busting behavior of the header is hacky and a bit unpredictable, but it’s also addressing a pretty important use case, so I don’t think we can do away with it without a viable replacement. Musing on the two options you present:

> making the "zoom" button a dropdown menu, allowing for more zoom levels

I think we could do this more intuitively by having separate “zoom in” and “zoom out” buttons next to each other in the header, but we’d also need to change their behavior to _also_ zoom the iframe (a quick spike in the dev tools suggests that this would work fine using `transform`). Perhaps that would be the desired behavior for zooming in general?

> a "presentation mode" where the header bar is just hidden from view

This is an intriguing idea. Would “presentation mode” also increase the zoom? How would one get the header back / exit presentation mode?

iamnickvolpe commented:
Yeah, totally agree about needing a viable replacement. To be honest, It wasn't until I was fixing my tests until I realized, "Wait, I wonder if the sizing was set like that for a reason." And sure enough, I found that old PR. :)

I'll tell you what. Let me adjust this to at least retain the current functionality, and then we'll work from there. It might not be quite as nice, but I can get something passable enough to get this in.

Then, I'd love to try messing with these two other potential features, because I think both could be pretty valuable regardless.

100% agree with the zoom buttons. A little plus and minus would be awesome. What about something like this?

<img width="456" alt="screenshot 2017-11-12 18 51 23" src="https://user-images.githubusercontent.com/3432105/32704872-9866dcac-c7da-11e7-9499-a23b92576e49.png">

As for "Presentation Mode" I was thinking that it wouldn't have to be anything more complicated than simply hiding the header bar. I feel like this wouldn't be a "saved" option and it would always show on load.

<img width="1128" alt="screenshot 2017-11-12 18 47 56" src="https://user-images.githubusercontent.com/3432105/32704891-cb46a760-c7da-11e7-9ff9-e71fc78fc206.png">

iamnickvolpe commented:
Er, I think that carat is backwards, but you get the gist.

iamnickvolpe commented:
OK, everything's in ems now. Even the shadows! So absolutely nothing changes on zoom. @outoftime let me know what you think of this one.

outoftime commented on src/css/application.css:
```diff
@@ -113,13 +113,15 @@ body {
   display: flex;
   flex: 0 0 auto;
-  font-size: 2vh;
+  font-family: sans-serif;
```
What’s the thinking here? I would expect this to be one of the project fonts, or nothing at all?

iamnickvolpe replied:
It was part of my original mockup. In general, it's a little wider and easier to read.

![21slczmmuh](https://user-images.githubusercontent.com/3432105/32761626-5cc1dff6-c8c2-11e7-8b84-c9b3c7fbaa6c.gif)

outoftime commented on src/css/application.css:
```diff
@@ -131,67 +133,87 @@ body {
 
 .top-bar__avatar {
-  height: 2em;
-  border-radius: 50%;
```
Hmm, honestly I think the circle looks quite a bit nicer…

iamnickvolpe replied:
I agree the circle is trendy and looks good, but 2 reasons I changed it. 1, Github uses squares and 2, it matches the other buttons and general style. Happy to go back to circle if you really feel strongly about it though.

outoftime replied:
I’m into the way the rounded corners look.

outoftime commented on src/css/application.css:
```diff
@@ -131,67 +133,87 @@ body {
+  box-shadow: inset 0 0.08em 0.2em 0 rgba(0, 0, 0, 0.25);
+}
+
+.top-bar__menu-button_projects .top-bar__menu {
```
No nesting in BEM!!

iamnickvolpe replied:
Should it be `menu-button--projects`? How would you do it?

outoftime commented on src/css/application.css:
```diff
@@ -131,67 +133,87 @@ body {
+  right: 0;
   position: absolute;
-  top: 100%;
+  top: 3.4em;
```
Where does this number come from? Does 100% not work?

iamnickvolpe replied:
I wanted the menus to line up with the bar instead of the button itself. It felt a little crowded and out of place to me.

![o5zp8up6h6](https://user-images.githubusercontent.com/3432105/32761865-cce0b1e4-c8c3-11e7-9c99-8d0269e01e6c.gif)

outoftime commented on src/css/application.css:
```diff
@@ -527,7 +547,8 @@ body {
 
 .notification-list__notification_notice {
-  background-color: var(--color-green);
+  background-color: #00b8ff;
```
Same as in other PR, please put all colors in variables!

outoftime commented on src/css/application.css:
```diff
@@ -527,7 +547,8 @@ body {
 .notification-list__notification_notice {
-  background-color: var(--color-green);
+  background-color: #00b8ff;
+  color: #fff;
```
This can just be `white` (no need for a custom variable)

iamnickvolpe commented:
OK @outoftime Try this out.

I made the updates you requested. A couple of things to net out on though:

- I decided to try a .2em border-radius on all the buttons, menus, and the avatar. Let me know what you think. (see Slack)

- And while I also like circle avatars in general, I think I would prefer to match Github, especially being that their default avatars are pixel-y characters that our circles tend to cut off. 

<img width="137" alt="screenshot 2017-11-13 23 56 18" src="https://user-images.githubusercontent.com/3432105/32763552-67ceb106-c8ce-11e7-833b-f3a86fdd5062.png">

- I still think sans-serif is a more readable font to use for interface elements, especially with bold text.

- Make sense what I'm saying about the menus? Posted a screenshot.

outoftime commented on src/components/TopBar/createMenu.jsx:
```diff
@@ -69,7 +70,9 @@ export default function createMenu({
-            className="top-bar__menu"
+            className={classnames('top-bar__menu',
+              menuClass,
+            )}
```
Not a huge deal but I feel like this could all go on one line?

iamnickvolpe replied:
Done.

outoftime commented on src/css/application.css:
```diff
@@ -113,13 +115,14 @@ body {
   font-weight: bold;
-  align-items: center;
-  padding: 0 1em;
+  padding: 1vh;
```
Shouldn’t this be `1vw 1vh`?

iamnickvolpe replied:
Actually it works just as well in ems. Changed to ems.

outoftime commented on src/css/application.css:
```diff
@@ -131,73 +134,107 @@ body {
+  margin: 0 0.4em 0 0.2em;
+  min-width: 1.6em;
+  transform: scale(1.8);
+  width: 1.6em;
```
Does this need both a `width` and a `min-width`?

iamnickvolpe replied:
Done. Handled differently.

outoftime commented on src/css/application.css:
```diff
@@ -131,73 +134,107 @@ body {
+.top-bar__logo-container {
+  margin: 0 0.4em 0 0.2em;
+  min-width: 1.6em;
+  transform: scale(1.8);
```
Bringing in `scale` seems like a weird move here, particularly given it’s an SVG and we can size it however we want!

Seems like the main hurdle here is that we want Pop to overflow the top/bottom padding that the rest of the menu items are subject to. A couple of alternative ideas:

1. Make Pop the size we want it using the `width` property, then position it relative and add a negative `top` to offset the padding
2. Instead of giving padding to the overall top bar, just give it an explicit `height`; the use of `align-items: center` will still keep everything looking nice. And we can again just give Pop the size we want

I’m strongly partial to Option 2, because it solves another problem, which I will mention below!

iamnickvolpe replied:
Option 2 is wayyy better than what I was doing. Done.

outoftime commented on src/css/application.css:
```diff
@@ -131,73 +134,107 @@ body {
-  margin-left: -0.25em;
+  margin: 0 1em 0 0.5em;
+  min-width: 6.5em;
+  width: 6.5em;
```
Wondering about the `min-width` and `width` together again?

iamnickvolpe replied:
Done. Handled differently.

outoftime commented on src/css/application.css:
```diff
@@ -131,73 +134,107 @@ body {
 .top-bar__avatar {
-  height: 2em;
-  border-radius: 50%;
+  border-radius: 0.2em;
```
Can we extract this radius into a var, since we’re using it in a bunch of places?

iamnickvolpe replied:
Done.

outoftime commented on src/css/application.css:
```diff
@@ -131,73 +134,107 @@ body {
-  border-color: transparent;
-  border-style: solid;
-  border-width: 0 1px;
+  align-self: center;
```
We don’t need this do we? Since the flex container has `align-items: center`?

iamnickvolpe replied:
Done.

outoftime commented on src/css/application.css:
```diff
@@ -131,73 +134,107 @@ body {
+  box-shadow: 0 0 20px rgba(0, 0, 0, 0.25);
   position: absolute;
-  top: 100%;
+  top: 3.4em;
```
Another advantage of using an explicit height for the top bar instead of padding is that we could go back to 100% height here. That’s very appealing because it means we don’t have the knowledge of “what is the outer height of the top bar” spread over several places in the stylesheet.

iamnickvolpe replied:
Still doesn't quite work at 100% if you want the menus to sit below the bar, not directly below the button. I'm not partial to the way this looks.

<img width="492" alt="screenshot 2017-11-16 20 33 09" src="https://user-images.githubusercontent.com/3432105/32924961-61a0aab4-cb0d-11e7-9e10-9b97562453fd.png">

outoftime commented on src/css/application.css:
```diff
@@ -131,73 +134,107 @@ body {
+.top-bar__menu_current-user,
+.top-bar__menu_hamburger {
+  right: 0;
+}
```
I’m a big proponent of just creating blocks/elements that straightforwardly describe the style that’s being applied, rather than naming everything semantically. You’ll find some semantic block/element names in Popcode but mostly for top-level layout concerns that don’t really have any scope for reuse.

Non-semantic block/element names are nice to work with because then the view templates can just straightforwardly describe how they’d like to look.

So: how about `top-bar__menu_left` and `top-bar__menu_right`?

iamnickvolpe replied:
Done.

iamnickvolpe commented:
- Merged latest
- Addressed reviews
- Added subtle shadow to top bar.

Latest design:
<img width="1324" alt="screenshot 2017-11-16 20 53 03" src="https://user-images.githubusercontent.com/3432105/32925489-3223e99c-cb10-11e7-8fc8-06ca03a50cf5.png">

outoftime commented on src/css/application.css:
```diff
@@ -68,6 +68,9 @@
   --color-purple: #f0f;
+  --color-low-contrast-blue: #baecff;
+  --color-blue: #00b8ff;
+  --rounded-corners: 0.2em;
```
How about `--size-rounded-corners` to maintain the naming convention established by `--color-` prefixed variables above?

outoftime commented on src/css/application.css:
```diff
@@ -133,72 +139,96 @@ body {
+  box-shadow: 0 0 20px rgba(0, 0, 0, 0.25);
   position: absolute;
-  top: 100%;
+  top: 3.4em;
```
So it makes sense why this can’t be 100%—I had forgotten that it is a child of the button and not the `TopBar` directly— but I still don’t love duplicating the knowledge of how tall overall the top bar is going to be. Can we extract that information into variables and/or use a `calc` statement to make more explicit how we get to `3.4em` here?

iamnickvolpe commented:
@outoftime Made the changes you requested. I think my math is right there. It **does** seem to work. Let me know what you think.
